### PR TITLE
librbd: tweak deep-copy to avoid creating last snapshot until sync is complete

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman there are theuthology failures [1]. I have not checked all but the most (all?) seemed to fail with:
```
2020-02-10T00:50:16.972 INFO:tasks.workunit.cluster1.client.mirror.smithi072.stderr:+ rbd --cluster cluster1 mirror pool enable --site-name cluster1 mirror pool
2020-02-10T00:50:17.001 INFO:tasks.workunit.cluster1.client.mirror.smithi072.stderr:rbd: failed to set site name(13) Permission denied
```
[1] http://pulpito.ceph.com/trociny-2020-02-09_17:47:19-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by trociny** [comment]
There are also strange permissions.sh test failures [1] (not sure related) in rbd_support, when `rbd task add remove volumes/foo` failed with "image note found", although the image create command had been reported to succeed:
```
2020-02-10T09:15:47.999 INFO:tasks.workunit.client.0.smithi080.stderr:+ rbd -k /tmp/tmp.Td9UCyRb6I --id volumes create --image-format 2 --image-feature layering,exclusive-lock,object-map,fast-diff -s 1 volumes/foo
2020-02-10T09:15:48.261 INFO:tasks.workunit.client.0.smithi080.stderr:+ ceph -k /tmp/tmp.Td9UCyRb6I --id volumes rbd task add remove volumes/foo
2020-02-10T09:15:48.334 INFO:tasks.ceph.mgr.x.smithi080.stderr:2020-02-10T09:15:48.322+0000 7fb1fc666700 -1 mgr.server reply reply (2) No such file or directory [errno 2] RBD image not found (Image volumes/foo does not exist)
2020-02-10T09:15:48.337 INFO:tasks.workunit.client.0.smithi080.stderr:Error ENOENT: [errno 2] RBD image not found (Image volumes/foo does not exist)
```
[1] http://pulpito.ceph.com/trociny-2020-02-10_08:57:07-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
Looks like it's happening in the master branch as well [1]. I don't think anything RBD merged recently so this could be a broken change to teuthology / the QA suite.

[1] http://pulpito.ceph.com/teuthology-2020-02-02_02:01:03-rbd-master-distro-basic-smithi/

**Comment by dillaman** [comment]
It was caused by [1] -- CEPH_ARGS parsing was broken back in Mimic so now that it's working again, the test is using the wrong credentials. I'll open a ticket and fix it.

[1] https://github.com/ceph/ceph/pull/32830

**Comment by trociny** [comment]
> It was caused by [1] -- CEPH_ARGS parsing was broken back in Mimic so now that it's working again, the test is using the wrong credentials.

@dillaman  To make mirror functional tests work locally (`RBD_MIRROR_USE_EXISTING_CLUSTER` is not set), additionally to your #33187 I also had to add `CEPH_ARGS=''` before `rbd mirror image|pool status` commands because they contact mgr for rbd-mirror daemon service status. And I had to override CEPH_ARGS in `daemon_asok_file` and `daemon_pid_file`, when calling `ceph-conf`, because the `--id` param from environment overrode `--name` param from the command line. And also in `start_mirror` helper (when starting rbd-mirror daemon), otherwise the id from env ("mirror") overrode the id from the command line ("mirror.{instance}") and did not read the instance specific setting in ceph.conf. This last case makes me think that  #32830 might not be entirely correct, because usually command line settings have the highest priority.

So I would prefer if in  #32830 it did not override args that already existed (obtained from the command line) and only added those that were new. It would not fix the case like we have with ceph-conf though (when `--id` from env overrides `--name` setting from the command line) but for this case it is not so obvious if the current behavior is wrong and should be fixed.

**Comment by dillaman** [comment]
#32830 just restores the original behaviour that was broken. The issue w/ the `ceph-conf` calls is that it's using `--name` when `CEPH_ARGS` has `--id` (I am also playing w/ this locally now). If it's not pulling the CLI overrides, however, that would be a bug and I'll open a new ticket for it. Thx.

**Comment by dillaman** [comment]
... ah yes, the environment is definitely overriding the environment args since we used to collect the env args first and then apply the CLI args.

**Comment by dillaman** [comment]
A tweak for `global_pre_init` is in #33213

**Comment by dillaman** [comment]
... and another fix for `CEPH_ARGS` processing in #33219

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2020-02-11_18:16:57-rbd-wip-jd-testing-distro-basic-smithi/

File: src/librbd/api/Mirror.cc [hunk 1]
```diff
@@ -1097,6 +1098,33 @@ int Mirror<I>::mode_set(librados::IoCtx& io_ctx,
...
 
+template <typename I>
+int Mirror<I>::uuid_get(librados::IoCtx& io_ctx, std::string* mirror_uuid) {
+  CephContext *cct = reinterpret_cast<CephContext *>(io_ctx.cct());
+  ldout(cct, 20) << dendl;
+
+  C_SaferCond ctx;
+  uuid_get(io_ctx, mirror_uuid, &ctx);
+  int r = ctx.wait();
+  if (r < 0 && r != -ENOENT) {
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: nit: shouldn't we clear `mirror_uuid` when we got `-ENOENT` and return 0 insted? Just thinking about the case, when a caller supplies a non-empty `mirror_uuid`, internaly `uuid_get` gets `-ENOENT` but returns 0 to the caller, so it is valid her to assume that `mirror_uuid` is overwritten with the valid content.

File: src/cls/rbd/cls_rbd_types.cc [hunk 2]
```diff
@@ -753,13 +754,15 @@ void MirrorNonPrimarySnapshotNamespace::decode(bufferlist::const_iterator& it) {
...
   decode(last_copied_object_number, it);
+  decode(snap_seqs, it);
 }
 
 void MirrorNonPrimarySnapshotNamespace::dump(Formatter *f) const {
   f->dump_string("primary_mirror_uuid", primary_mirror_uuid);
   f->dump_unsigned("primary_snap_id", primary_snap_id);
   f->dump_bool("copied", copied);
   f->dump_unsigned("last_copied_object_number", last_copied_object_number);
+  f->dump_stream("snap_seqs") << snap_seqs;
```
Comment (trociny) [inline src/cls/rbd/cls_rbd_types.cc]: nit: don't we want to update `operator<<(std::ostream&, const MirrorNonPrimarySnapshotNamespace&)` too?

File: src/librbd/api/Mirror.cc (hunk 1 above)
Comment (dillaman) [reply src/librbd/api/Mirror.cc]: I broke this during an internal refactor to add the async version. It used to just avoid printing the error message on `-ENOENT` but still return the error code. Fixed.

File: src/cls/rbd/cls_rbd_types.cc (hunk 2 above)
Comment (dillaman) [reply src/cls/rbd/cls_rbd_types.cc]: Added

**Comment by dillaman** [comment]
(dropped DNM commits from other PRs)

**Pull request merged**